crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,297 @@
+                            update: my_proposal.clone(),
+                        }),
+                    ),
+                };
+                response_sender.send(response);
+                Ok(())
+            },
+            _ => {
+                bail!("unexpected rpc: {}", msg.name());
+            },
+        }
+    }
+
+    /// Triggered once the `certified_update_producer` produced a quorum-certified update.
+    pub fn process_quorum_certified_update(&mut self, update: QuorumCertifiedUpdate) -> Result<()> {
+        let state = self
+            .states_by_issuer
+            .entry(update.observed.issuer.clone())
+            .or_default();
+        match &state.consensus_state {
+            ConsensusState::InProgress { my_proposal, .. } => {
+                //TODO: counters
+                state.consensus_state = ConsensusState::Finished {
+                    my_proposal: my_proposal.clone(),
+                    quorum_certified: update,
+                };
+                self.update_vtxn_pool()?;
+                Ok(())
+            },
+            _ => Err(anyhow!(
+                "qc update not expected for issuer {:?} in state {}",
+                update.observed.issuer,
+                state.consensus_state.name()
+            )),
+        }
+    }
+
+    fn update_vtxn_pool(&mut self) -> Result<()> {
+        let updates: BTreeMap<Issuer, QuorumCertifiedUpdate> = self
+            .states_by_issuer
+            .iter()
+            .filter_map(
+                |(issuer, per_provider_state)| match &per_provider_state.consensus_state {
+                    ConsensusState::Finished {
+                        quorum_certified, ..
+                    } => Some((issuer.clone(), quorum_certified.clone())),
+                    _ => None,
+                },
+            )
+            .collect();
+        let txn = ValidatorTransaction::ObservedJWKsUpdates { updates };
+        self.vtxn_pool_write_cli.put(Some(Arc::new(txn)));
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct AbortHandleWrapper {
+    handle: Option<AbortHandle>,
+}
+
+impl AbortHandleWrapper {
+    pub fn new(handle: AbortHandle) -> Self {
+        Self {
+            handle: Some(handle),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn dummy() -> Self {
+        Self { handle: None }
+    }
+}
+
+impl Drop for AbortHandleWrapper {
+    fn drop(&mut self) {
+        let AbortHandleWrapper { handle } = self;
+        if let Some(handle) = handle {
+            handle.abort();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum ConsensusState {
+    NotStarted,
+    InProgress {
+        my_proposal: ObservedUpdate,
+        abort_handle_wrapper: AbortHandleWrapper,
+    },
+    Finished {
+        my_proposal: ObservedUpdate,
+        quorum_certified: QuorumCertifiedUpdate,
+    },
+}
+
+impl PartialEq for ConsensusState {
+    fn eq(&self, other: &Self) -> bool {
```

Comment:
Here it is checking ObservedUpdate which includes the author, is this the right behavior?

Reply:
Iirc it's by design, and rhe `ConsensusState::eq` is so far only used to assert states in tests